### PR TITLE
CUse: include missing class forward declarations

### PR DESCRIPTION
**Pull request by kboronski-ant**
* Organizes forward declarations into a single block in design internal headers
* Simplifies visiting `AstClassRefDType`s and prevents marking types as implementation-only (it's unnecessary and unused, because implmenetations get the `__Syms` header file. It also resulted in skipping them on later visits, where they could've been deemed necessary to be forward-delared in interfaces, ie. headers.)
* Adds `CUse`s for `AstClassRefDType`s that are pointed to by typeref nodes that are not `AstClassRefDType`
* Ignores types referenced in function bodies only.

### CODE REVIEW

File: src/V3EmitCHeaders.cpp [hunk 1]
```diff
@@ -22,6 +22,7 @@
 #include "V3Global.h"
 
 #include <algorithm>
+#include <unordered_set>
```
Comment (wsnyder) [inline src/V3EmitCHeaders.cpp:25]:
```suggestion
#include <set>
```

unordered set will be slower here

File: src/V3EmitCBase.cpp [hunk 2]
```diff
@@ -224,21 +224,12 @@ void EmitCBaseVisitorConst::emitVarDecl(const AstVar* nodep, bool asRef) {
...
-                }
-                if (usep->useType().isFwdClass()) {
-                    puts("class " + prefixNameProtect(usep) + ";\n");
-                }
-                nl = "\n";
-            }
-        }
-    }
-    puts(nl);
+    bool nl;
```
Comment (wsnyder) [inline src/V3EmitCBase.cpp:227]:
```suggestion
    bool nl = false;
```

File: src/V3CUse.cpp [hunk 3]
```diff
@@ -61,12 +65,31 @@ class CUseVisitor final : public VNVisitor {
...
+            m_dtypesImplOnly = true;
+
+            iterateAndNextNull(nodep->initsp());
+            iterateAndNextNull(nodep->stmtsp());
+            iterateAndNextNull(nodep->finalsp());
+        }
+    }
+    void visit(AstCReturn* nodep) override {
+        if (nodep->user1SetOnce()) return;  // Process once
+
```
Comment (wsnyder) [inline src/V3CUse.cpp:86]:
```suggestion
```

File: src/V3CUse.cpp [hunk 4]
```diff
@@ -61,12 +65,31 @@ class CUseVisitor final : public VNVisitor {
...
-        if (!m_impOnly) addNewUse(nodep, VUseType::INT_FWD_CLASS, nodep->classp()->name());
-        // Need to include extends() when we implement, but no need for pointers to know
-        VL_RESTORER(m_impOnly);
+        addNewUse(nodep, VUseType::INT_FWD_CLASS, nodep->classp()->name());
+    }
+    void visit(AstCFunc* nodep) override {
+        if (nodep->user1SetOnce()) return;  // Process once
+
+        iterateAndNextNull(nodep->argsp());
+
```
Comment (wsnyder) [inline src/V3CUse.cpp:74]:
```suggestion
        iterateAndNextNull(nodep->argsp());
```

File: src/V3EmitCHeaders.cpp (hunk 1 above)
Comment (kboronski-ant) [reply src/V3EmitCHeaders.cpp]: Ok, I changed it to `set`.

**Comment by wsnyder** [comment]
Is this still a draft?  I'm ok with your squash merging if you think it's done and tests pass.

**Pull request merged**